### PR TITLE
[StructuralMechanicsApplications] Adding explicit improvements (new nodal variables + MPC)

### PR DESCRIPTION
**Pull request by loumalouomega**
This PR adds:
- New nodal variables to reduce computational cost of computing damping as part of RHS
- Adding explicit MPC computation
- Test for explicit MPC

### CODE REVIEW

File: applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_central_differences_scheme.hpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_central_differences_scheme.hpp]: do not use C arrays(use std::array if you need this functionality)

File: applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_central_differences_scheme.hpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_central_differences_scheme.hpp:509]: this is not as simple as this... you cannot lump the Inertia tensor, unless you work in the principal system of coordinates

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp:99]: what are u doing with the ReformDof ?

please take into accoutn that using dofs is quite slow, so try to minimize their usage in an explicit context

to be more clear, they are not the dofs per se to be slow. it is the access to the dofs, so the call to

node->pGetdof

if you have a doflist already initialized you can well work with that

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp]: ?critical?
this is a performance killer!!

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp]: care to get a reference and not a copy here

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp]: same

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp]: i advocate for adding a function Apply to the MasterSlaveconstraints so that all of the ops above can be done in there

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp:495]: I only have one request, this function has nothing to do with explicit, it fact it is also needed in many other places (e.g. the eigen-postprocessing also needs this)
=> please put it to an independent file in the core e.g. `ReconstructConstraintSlaveSolutionUtility` or so
(this way it could also be used from python, which might come in handy at some point)

FYI @adityaghantasala

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp]: const, the Constraints already only accept const ProcessInfo (unlike Elements and Conditions which need #2993 for this)

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp:605]: :+1:

File: applications/StructuralMechanicsApplication/structural_mechanics_application_variables.h
Comment (philbucher) [inline applications/StructuralMechanicsApplication/structural_mechanics_application_variables.h]: can you call it `NODAL_ROTATION_DAMPING` in order to be consistent? All other variables that have "rotation" in the name are using "ROTATION"

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp:285]: is it much effort to refactor?

**Review by philbucher — CHANGES_REQUESTED** [review]
Thanks, really cool!
@KlausBSautter once again you were saved by @loumalouomega , he implemented what I told you the other day what you should implement ... :+1: 

one minor refactor request, otherwise looks good (please take into account the comments of @RiccardoRossi )

**Comment by KlausBSautter** [comment]
Nice, @philbucher is right, we were just talking about this :)

File: applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_central_differences_scheme.hpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_central_differences_scheme.hpp]: In my defense this was already implemented, but you are right

File: applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_central_differences_scheme.hpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_central_differences_scheme.hpp:509]: That hypothesis was already there, ask @KlausBSautter for details

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp:99]: The thing is that this member variable is not used, I just changed the way the constructor works

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp:99]: If I remove the variable that would be a behavior change. For a future PR

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp:99]: In conclussion: It was already there, is not used, I just move the way the constructor works

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp]: I know, the alternative is to construct a dof set in the Initialize with the slave dofs

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp]: True

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp:285]: Well, no, but better not create  a PR with thousand of changes

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp:495]: @philbucher @loumalouomega i really think that we need support for this from the MasterSlaveConstraints (the famous apply method...)

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp]: you still have a very severe bottleneck there. Consider doing something like

~~~cpp
double& value = slave_equation_dofs[i]->GetSolutionStepValue();
#pragma omp atomic
value *= 0.0;
~~~

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp]: Nice trick I will try

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp:495]: I created the Apply method I will put it in this PR and in a standalone PR

File: applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_central_differences_scheme.hpp
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_central_differences_scheme.hpp:509]: @KlausBSautter this is only ok if you update the inertia when you rotate the beam local axes, and addionally in the hypothesis of working in the principle system of intertia of the beam. otherwise you are doing a major error with this

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp]: It works, I din't know *= was compatible with atomic. I thought was only -= and +=

**Comment by loumalouomega** [comment]
@philbucher and @RiccardoRossi suggestions applied

File: applications/StructuralMechanicsApplication/custom_utilities/constraint_utilities.h
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_utilities/constraint_utilities.h]: please move it to the core, it is certainly general enough

Thanks btw, looks really good

File: applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_central_differences_scheme.hpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_central_differences_scheme.hpp]: as a performance improvement i would use the position suggestion as it is done here. 

https://github.com/KratosMultiphysics/Kratos/blob/master/applications/FluidDynamicsApplication/custom_elements/fluid_element.cpp#L286

and 

https://github.com/KratosMultiphysics/Kratos/blob/master/applications/FluidDynamicsApplication/custom_elements/fluid_element.cpp#L291

this will (greatly) speedup the getdof search

File: kratos/includes/linear_master_slave_constraint.h
Comment (RiccardoRossi) [inline kratos/includes/linear_master_slave_constraint.h]: I would suggest to avoid the allocation of the slave_dofs_values vector.

this could be done by doing

~~~cpp
Vector master_dofs_values(mMasterDofsVector.size());

        for (IndexType i = 0; i < mMasterDofsVector.size(); ++i) {
            master_dofs_values[i] = mMasterDofsVector[i]->GetSolutionStepValue();
        }

        for (IndexType i = 0; i < mSlaveDofsVector.size(); ++i) {
            double aux = mConstantVector[i];
            for(IndexType j=0; j<mRelationMatrix.size2(); ++j)
                   aux += mRelationmatrix(i,j)*slave_dofs_values[j];
                 
            #pragma omp atomic
            mSlaveDofsVector[i]->GetSolutionStepValue() += aux;
        }
~~~

Comment (loumalouomega) [reply kratos/includes/linear_master_slave_constraint.h]: It makes sense

File: applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_central_differences_scheme.hpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_central_differences_scheme.hpp]: Ok

File: kratos/utilities/constraint_utilities.h
Comment (philbucher) [inline kratos/utilities/constraint_utilities.h]: take care

File: kratos/includes/linear_master_slave_constraint.h
Comment (philbucher) [inline kratos/includes/linear_master_slave_constraint.h]:
```suggestion
            mSlaveDofsVector[i]->GetSolutionStepValue() *= 0.0;
```

**Comment by loumalouomega** [comment]
Done

File: kratos/utilities/constraint_utilities.h
Comment (loumalouomega) [reply kratos/utilities/constraint_utilities.h]: True

File: kratos/utilities/constraint_utilities.cpp
Comment (philbucher) [inline kratos/utilities/constraint_utilities.cpp:24]: Here you called it `ResetSlaveDofs` which I think sounds a lot more natural
@loumalouomega @RiccardoRossi @KratosMultiphysics/technical-committee are we sure we want to stick with `ResetSlave` in #4163 

Once merged there will most likely be no way back ...

File: applications/StructuralMechanicsApplication/structural_mechanics_application_variables.cpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/structural_mechanics_application_variables.cpp:131]: why is `NODAL_ROTATION_DAMPING` an array and `NODAL_DISPLACEMENT_DAMPING` a double?

**Review by philbucher — COMMENTED** [review]
one question remaining, otherwise I will approve after #4163

File: applications/StructuralMechanicsApplication/structural_mechanics_application_variables.cpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/structural_mechanics_application_variables.cpp:131]: There is not a mass per each direction. In fact the rotational should be a tensor not an array as @RiccardoRossi says

Comment (philbucher) [reply applications/StructuralMechanicsApplication/structural_mechanics_application_variables.cpp:131]: ok
@KlausBSautter to be discussed

Comment (KlausBSautter) [reply applications/StructuralMechanicsApplication/structural_mechanics_application_variables.cpp:131]: @philbucher we can discuss this tmr at the chair

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp:496]: this argument is not used

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp:496]: Just in case

File: kratos/utilities/constraint_utilities.cpp
Comment (RiccardoRossi) [reply kratos/utilities/constraint_utilities.cpp:24]: #4163 already has the name changed. allright to me

File: applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_central_differences_scheme.hpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_central_differences_scheme.hpp]: have this out of the loop, this is the expensive function...

actually you may even want to store disppos and rotpos as class members...

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_central_differences_scheme.hpp]: I thought was something like the EquationId, so is common in the variable list?, OK, updating

File: kratos/utilities/constraint_utilities.cpp
Comment (philbucher) [reply kratos/utilities/constraint_utilities.cpp:24]: yes is resolved

File: applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_central_differences_scheme.hpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_central_differences_scheme.hpp]: Done

File: applications/StructuralMechanicsApplication/structural_mechanics_application_variables.cpp
Comment (philbucher) [reply applications/StructuralMechanicsApplication/structural_mechanics_application_variables.cpp:131]: on monday, I am not there tmr :)

**Review by philbucher — APPROVED** [review]

**Pull request merged**
